Improve include time of Mojo::Base by extracting monkey_patch

okurz opened the pull request:
### Motivation
Commit 7e9a2ad introduced a "require Mojo::Util" causing a significant chain of further dependencies being pulled in which IMHO should be avoided for the very base module which is in particular being advertised as useful for just enabling static code checks and common import checks.

### Summary
This commit moves out the function "monkey_patch" into its own module to break or prevent the circular dependency between Mojo::Base and Mojo::Util.

With that the import of Mojo::Base is more efficient, `time perl -e 'use Mojo::Base`
on my system reduced from 224±12.08 ms to 52.0±2.3 ms which I consider a considerable improvement for Mojo::Base which is used as a baseclass in many cases.

guest20 commented on lib/Mojo/Util.pm:
```diff
@@ -75,6 +75,13 @@ our @EXPORT_OK = (
+
+# redirecting to MonkeyPatch::monkey_patch for backward compatibility
+    no warnings 'redefine';
+sub monkey_patch {
```
You already import it, why re-define it here?

perlpunk replied:
If it is imported via `use` then it's also defined in the Mojo::Util namespace, so you can export it.

jberger replied:
you should be able to re-export symbols, I don't think that should be a problem

okurz replied:
will do that then :)

okurz commented on lib/Mojo/Base.pm:
```diff
@@ -10,8 +10,7 @@ use mro;
 
-# Defer to runtime so Mojo::Util can use "-strict"
-require Mojo::Util;
+use Mojo::MonkeyPatch qw(monkey_patch);
```
ok, so `require Mojo::MonkeyPatch` then?

Grinnz replied:
`use Mojo::MonkeyPatch ();` is sufficient, require was only used to allow the circular dependency.

okurz replied:
Understood, done

okurz commented:
* Addressed CI failure "Undefined subroutine &Mojo::Util::class_to_path" by moving class_to_path to Mojo::Base and adjusting the rest of the repository accordingly, re-exporting from Mojo::Util
* Removed the re-implementation of monkey_patch in Mojo::Util, directly re-exporting
* Replaced "use" by "require" as per https://github.com/mojolicious/mojo/pull/2173#discussion_r1553828119
* Fixed tests
* Added POD in Mojo::MonkeyPatch

Grinnz commented:
> * Addressed CI failure "Undefined subroutine &Mojo::Util::class_to_path" by moving class_to_path to Mojo::Base and adjusting the rest of the repository accordingly, re-exporting from Mojo::Util

This should not be done. This creates a method in all classes using Mojo::Base as a base.

Grinnz commented:
Additionally, if Mojo::Util re-exports class_to_path no other adjustment should be necessary.

okurz commented:
> > * Addressed CI failure "Undefined subroutine &Mojo::Util::class_to_path" by moving class_to_path to Mojo::Base and adjusting the rest of the repository accordingly, re-exporting from Mojo::Util
> 
> This should not be done. This creates a method in all classes using Mojo::Base as a base.

ok, do you have a preferred alternative? Move to yet another new module with a single function, combine with Mojo::MonkeyPatch in BaseUtils (???), re-implement the one-liner?

okurz commented:
> > * Addressed CI failure "Undefined subroutine &Mojo::Util::class_to_path" by moving class_to_path to Mojo::Base and adjusting the rest of the repository accordingly, re-exporting from Mojo::Util
> 
> This should not be done. This creates a method in all classes using Mojo::Base as a base.

done, updated

Grinnz commented:
BaseUtil seems reasonable and self-explanatory to me, but there may be other opinions. As long as Mojo::Util imports and re-exports the functions from there correctly, there should be no changes needed to other modules.

okurz commented:
> BaseUtil seems reasonable and self-explanatory to me, but there may be other opinions. As long as Mojo::Util imports and re-exports the functions from there correctly, there should be no changes needed to other modules.

Sure, let's see how that looks. Added a commit to combine monkey_patch and class_to_path in Mojo::BaseUtil and reverted changes in other modules to use still the functions as exported from Mojo::Util.

Grinnz commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,71 @@
+package Mojo::BaseUtil;
+use Mojo::Base -strict;
```
This will either need to be replaced with the same pragmas as Mojo::Base uses, or Mojo::Base will need to require Mojo::BaseUtil as it did Mojo::Util previously, because of the circular require.

okurz replied:
Currently I used in Base `use Mojo::BaseUtil ();` and CI tests passed. IIUC that means we handle the circular require already. Or should I change it to require in Mojo::Base still? Can you elaborate?

jberger replied:
I think it makes sense for this to just `use strict; use warnings;` IMO. The whole point is to make it extremely clear that this package breaks circular deps.

okurz replied:
got it, done

Grinnz replied:
> Currently I used in Base `use Mojo::BaseUtil ();` and CI tests passed. IIUC that means we handle the circular require already. Or should I change it to require in Mojo::Base still? Can you elaborate?

This will still fail to use strict and warnings and the other pragmas which -strict is supposed to import, because of the circular require. Specifically, Mojo::BaseUtil is loaded during the compilation phase of Mojo::Base, and attempts to load Mojo::Base which has not finished loading and so silently ignores its import method.

Grinnz commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,71 @@
+use Exporter  qw(import);
+use Sub::Util qw(set_subname);
+
+our @EXPORT_OK = (qw(monkey_patch));
```
Missing class_to_path export

Grinnz commented:
IMO, t/mojo/util.t should remain unchanged, to ensure there is no change in compatibility for the re-exported functions. t/mojo/base_util.t may be unnecessary if so.

kraih commented:
I'd be interested in some real world benchmarks where this makes a difference.

poti1 commented:
I'm looking forwards to seeing this hopefully soon in the library :)

poti1 commented:
By simply copying over `monkey_patch` and the `set_subname` declarations from Mojo::Util to my cli tool (and removing use/require Mojo::Util), I saw these results in performance improvement:

BEFORE:
real    0m0.179s
user    0m0.101s
sys     0m0.052s

NOW:
real    0m0.045s
user    0m0.018s
sys     0m0.015s

When running on my phone's termux, the change of those values is really felt.

Grinnz commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,71 @@
+
+=head1 DESCRIPTION
+
+L<Mojo::BaseUtil> provides a C<class_to_path> and the C<monkey_patch> function for L<Mojo>.
```
It may be worth detailing here the purpose of this module: to provide functions for both Mojo::Base and Mojo::Util, so Mojo::Base does not have to load the rest of Mojo::Util.

okurz replied:
That's a good idea. I more or less included your sentence literally as I think it's already phrased in the best understandable way.

okurz commented:
* Updated POD for description
* Fixed one CI failure

okurz commented:
* Using only `strict+warnings` in BaseUtils to address https://github.com/mojolicious/mojo/pull/2173#discussion_r1594074777

kraih commented:
Please squash commits.

okurz commented:
> Please squash commits.

done

kraih commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,78 @@
+package Mojo::BaseUtil;
+
+# only using pure Perl as the only purpose of this module is to break a
```
Comment capitalisation. And our line length is 120 chars.

Grinnz commented on t/mojo/util.t:
```diff
@@ -6,9 +6,8 @@ use lib curfile->sibling('lib')->to_string;
 use Test::More;
 use Mojo::ByteStream qw(b);
 use Mojo::DeprecationTest;
-use Sub::Util qw(subname);
```
The changes to this test file are inconsistent, and I think this test file could actually be left alone so as to make sure that the re-exports always remain functional.

Grinnz commented on lib/Mojo/Util.pm:
```diff
@@ -67,6 +67,7 @@ my $ENTITY_RE = qr/&(?:\#((?:[0-9]{1,7}|x[0-9a-fA-F]{1,6}));|(\w+[;=]?))/;
 # Encoding and pattern cache
 my (%ENCODING, %PATTERN);
 
+# re-exporting BaseUtil functions for backward compatibility
```
I would consider these re-exports a permanent functionality matter, not backwards compatibility - I don't see a reason to ever remove these from Mojo::Util. This comment isn't really needed

okurz commented:
Covered all last three open review comments

Grinnz reviewed (approved):
Looks good to me

jberger commented on lib/Mojo/Util.pm:
```diff
@@ -14,9 +14,9 @@ use IO::Poll qw(POLLIN POLLPRI);
+use Mojo::BaseUtil     qw(class_to_path monkey_patch);
 use Pod::Usage         qw(pod2usage);
 use Socket             qw(inet_pton AF_INET6 AF_INET);
-use Sub::Util          qw(set_subname);
```
Do you think anyone depends on Mojo::Util::set_subname? They probably shouldn't but they could

Grinnz replied:
I would not consider use of any subroutines imported without a documented interface as our responsibility.

kraih replied:
No documentation, no feature.

jberger commented on t/mojo/base_util.t:
```diff
@@ -0,0 +1,49 @@
+
+subtest 'monkey_patch' => sub {
+  {
+
```
nitpick but blank line?

okurz replied:
ok, done. Included is a second commit removing redundant blank lines in the added code as well as where I originally copied the test code from.

okurz replied:
Nope, perltidy actually demands the blank line :D Reverted the last change

jhthorsen commented:
I think the name Mojo::BaseUtil is a bit awkward. How to tell what is in which package? I can't come up with a killer name, but I'm thinking about Mojo::CoreUtil or Mojo::PureUtil, which might indicate that the functions in that package is without any *significant* dependencies? I guess Mojo::BaseUtil makes sense, if we only have functions that Mojo::Base depends on, but I wonder if it's a slippery slope...

Should any of the following functions be moved over? The reason I ask, is because maybe it makes sense to move class_to_file(), when class_to_path() was moved.

* camelize()
* class_to_file()
* decamelize()

Grinnz commented:
While that might make sense for code level consistency, I think it's beneficial to keep the module limited in scope to its stated purpose. All of these functions should continue to be used from Mojo::Util in normal cases. It can be viewed less as "moving" the functions from the user's POV because they continue to be re-exported.

jhthorsen commented:
Then I think that should be stated in the documentation, meaning that functions can be added and removed in the future to keep Mojo::Base quick to load.

Grinnz commented:
It could make sense to me to remove the function documentation from the new module (these would need to be added as trusted in t/pod_coverage.t), and document that it provides functions for use in Mojo::Base and shouldn't be depended on to remain compatible.

okurz commented:
> It could make sense to me to remove the function documentation from the new module

ok, fine for me. As there was already one "+1" I followed that suggestion and included that change

Also included is a second commit removing redundant blank lines in the added code as well as where I originally copied the test code from.

poti1 commented:
For my own cli script, I ended up doing this to avoid loading in thousands of subs right away:

```perl
sub monkey_patch {
    my ( $class, %patch ) = @_;
    require Sub::Util;
    for ( keys %patch ) {
        *{"${class}::$_"} =
          Sub::Util::set_subname( "${class}::$_", $patch{$_} );
    }
}
```

kraih commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,48 @@
+
+our @EXPORT_OK = (qw(class_to_path monkey_patch));
+
+
```
Nitpick: Extra newline.

kraih commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,48 @@
+L<Mojo::BaseUtil> provides a C<class_to_path> and the C<monkey_patch> function
+for L<Mojo>. The main purpose is to provide functions to both C<Mojo::Base>
+and C<Mojo::Util> so that C<Mojo::Base> does not have to load the rest of
+C<Mojo::Util>.
```
Nitpick: Not 120 char lines.

kraih commented on t/mojo/base_util.t:
```diff
@@ -0,0 +1,49 @@
+use Mojo::BaseUtil qw(class_to_path monkey_patch);
+
+
+subtest 'class_to_path' => sub {
```
Nitpick: Extra newline.

kraih commented on t/pod_coverage.t:
```diff
@@ -11,4 +11,7 @@ my @await = (
 );
 
-all_pod_coverage_ok({also_private => ['BUILD_DYNAMIC', @await, 'spurt']});
+# base utils only to be used in Mojo::Base and should not be used elsewhere
```
Nitpick: Comment capitalisation.

kraih replied:
Is that correct english? Sounds a little wrong to me.

okurz replied:
Reworded to 
> These are base utils only to be used in Mojo::Base and not elsewhere

kraih commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,46 @@
+
+=head1 DESCRIPTION
+
+L<Mojo::BaseUtil> provides a C<class_to_path> and the C<monkey_patch> function for L<Mojo>. The main purpose is to
```
That phrase sounds completely wrong, it's two functions and not for `Mojo`.

okurz replied:
Rephrased. Please check again.

kraih commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,46 @@
+
+=head1 NAME
+
+Mojo::BaseUtil - Common utility functions used in Mojo::Base, re-exported in Mojo::Util.
```
No punctuation for the package description.

kraih commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,45 @@
+
+=head1 DESCRIPTION
+
+L<Mojo::BaseUtil> provides functions to both C<Mojo::Base> and C<Mojo::Util> so that C<Mojo::Base> does not have to load
```
You've used `C<...>` instead of `L<...>` for the modules, otherwise it sounds good.

okurz replied:
Done

eduardoj commented:
Small typo in commit message and in pull request description: `strictures` shold be `structures`.

Grinnz commented on lib/Mojo/BaseUtil.pm:
```diff
@@ -0,0 +1,45 @@
+package Mojo::BaseUtil;
+
+# Only using pure Perl as the only purpose of this module is to break a circular dependency involving Mojo::Base
+use strict;
```
Minor detail, I think we should `use v5.16;` here to include the feature bundle as Mojo::Base -strict would. The rest of the things it imports are unlikely to be important here.

Grinnz replied:
or alternatively, add `use feature ':5.16';` to match Mojo::Base's declarations.

okurz replied:
So I now did the same as in lib/Mojo/Base.pm

```
use strict;
use warnings;
use feature ':5.16';
```

good like this?

okurz commented:
> Small typo in commit message and in pull request description: `strictures` shold be `structures`.

uh, no. I meant strictures as from https://metacpan.org/pod/strictures but maybe that word only makes sense when using that pod. So rephrased.

Grinnz reviewed (approved):
LGTM. don't know why perltidy test is failing

kraih commented:
> LGTM. don't know why perltidy test is failing

New perltidy release. Just rebase the PR on main and it will be fine.

okurz commented:
Rebased

Grinnz commented:
Thank you for persisting to tidy this up!
